### PR TITLE
Add BaseFragment and setup injection

### PR DESCRIPTION
**Pull request by anstaendig**
Issue: #12 

- Adds a BaseFragment and the needed setup for injection.
- Add Logging Interceptor
- Changed point of injection into lifecycle callback

### CODE REVIEW

File: app/src/main/java/anstaendig/com/architecturecomponents/injection/DatasourceModule.kt [hunk 1]
```diff
@@ -20,10 +20,13 @@ class DatasourceModule {
 
   @Provides
   @Singleton
-  fun provideOkHttpClient(): OkHttpClient {
-    val okHttpClient = OkHttpClient.Builder()
-    return okHttpClient.build()
-  }
+  fun provideOkHttpClient(loggingInterceptor: HttpLoggingInterceptor): OkHttpClient
+      = OkHttpClient.Builder().addInterceptor(loggingInterceptor).build()
```
Comment (jonyas) [inline app/src/main/java/anstaendig/com/architecturecomponents/injection/DatasourceModule.kt]: I know this is a example, but we should really think (if we add this to main module) to exclude this logs in release builds

File: app/src/main/java/anstaendig/com/architecturecomponents/ui/base/BaseActivity.kt [hunk 2]
```diff
@@ -3,38 +3,43 @@ package anstaendig.com.architecturecomponents.ui.base
...
+  // We have to hold a strong reference to the instance of LifecycleRegistry, otherwise GC will
+  // collect it and LifecycleObserver of LiveData won’t see the active observers from LifecycleOwner
+  // and our views won’t receive the updates.
+  @Suppress("LeakingThis")
+  private val lifecycleRegistry = LifecycleRegistry(this)
+  override fun getLifecycle() = lifecycleRegistry
 
   @CallSuper
   override fun onCreate(savedInstanceState: Bundle?) {
-    AndroidInjection.inject(this)
```
Comment (jonyas) [inline app/src/main/java/anstaendig/com/architecturecomponents/ui/base/BaseActivity.kt]: So now the injection is done with the app injector?

**Comment by anstaendig** [comment]
This is where I got a lot of the ideas from:
https://github.com/googlesamples/android-architecture-components

File: app/src/main/java/anstaendig/com/architecturecomponents/ui/base/BaseActivity.kt (hunk 2 above)
Comment (anstaendig) [inline app/src/main/java/anstaendig/com/architecturecomponents/ui/base/BaseActivity.kt]: Yes

File: app/src/main/java/anstaendig/com/architecturecomponents/injection/DatasourceModule.kt (hunk 1 above)
Comment (anstaendig) [inline app/src/main/java/anstaendig/com/architecturecomponents/injection/DatasourceModule.kt]: Good point

File: app/src/main/java/anstaendig/com/architecturecomponents/injection/AppInjector.kt [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+object AppInjector {
+
+  fun init(app: App) {
+    DaggerAppComponent
+        .builder()
+        .application(app)
+        .build()
+        .inject(app)
+    app.registerActivityLifecycleCallbacks(object : Application.ActivityLifecycleCallbacks {
+      override fun onActivityPaused(activity: Activity) {
```
Comment (jonyas) [inline app/src/main/java/anstaendig/com/architecturecomponents/injection/AppInjector.kt]: I hate the need to implement all methods when you implement a inner "interface". I think Kotlin awesomeness could do something about it. We should open a feature request :)

Comment (anstaendig) [inline app/src/main/java/anstaendig/com/architecturecomponents/injection/AppInjector.kt]: True

**Comment by jonyas** [comment]
Gonna read those fancy Dagger fragment injectors in the documentation, makes things super clean

**Pull request merged**